sql/src/main/java/io/crate/execution/engine/indexing/UpsertResultCollectors.java

```diff
@@ -0,0 +1,153 @@
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.execution.engine.indexing;
+
+import com.google.common.collect.ImmutableMap;
+import io.crate.data.Row;
+import io.crate.data.Row1;
+import io.crate.execution.dml.ShardResponse;
+import io.crate.execution.dml.upsert.ShardUpsertRequest;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.common.util.set.Sets;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.function.BinaryOperator;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collector;
+
+final class UpsertResultCollectors {
+
+    static Collector<ShardUpsertRequestAndResponse, UpsertResults, Iterable<Row>> newRowCountCollector() {
+        return new RowCountCollector();
+    }
+
+    static Collector<ShardUpsertRequestAndResponse, UpsertResults, Iterable<Row>> newSummaryCollector(DiscoveryNode localNode) {
+        return new SummaryCollector(ImmutableMap.of(
+            "id", new BytesRef(localNode.getId()),
+            "name", new BytesRef(localNode.getName())
+        ));
+    }
+
+    private static class RowCountCollector implements Collector<ShardUpsertRequestAndResponse, UpsertResults, Iterable<Row>> {
+
+        private final Object lock = new Object();
+
+        @Override
+        public Supplier<UpsertResults> supplier() {
+            return UpsertResults::new;
+        }
+
+        @Override
+        public BiConsumer<UpsertResults, ShardUpsertRequestAndResponse> accumulator() {
+            return this::processShardResponse;
+        }
+
+        @Override
+        public BinaryOperator<UpsertResults> combiner() {
+            return (i, o) -> {
+                synchronized (lock) {
+                    o.addResult(i.getSuccessRowCountForNoUri());
+                }
+                return o;
+            };
+        }
+
+        @Override
+        public Function<UpsertResults, Iterable<Row>> finisher() {
+            return r -> Collections.singletonList(new Row1(r.getSuccessRowCountForNoUri()));
+        }
+
+        @Override
+        public Set<Characteristics> characteristics() {
+            return Sets.newHashSet(Characteristics.CONCURRENT);
```

Comment:
Could use `Collections.singleton` - stock java instead of ES custom code and would also follow the recommendation of the `characteristics` description that this should be immutable